### PR TITLE
ENH Add support for running doctests in pyodide

### PR DESCRIPTION
**Pull request by hoodmane**
If the first line of the doctest has `# doctest: +RUN_IN_PYODIDE`, then we will run it in Pyodide once for each runtime available.

- [x] add a test
- [x] update changelog

### CODE REVIEW

File: pytest_pyodide/hook.py [hunk 1]
```diff
@@ -91,23 +92,42 @@ def pytest_configure(config):
...
-    if not hasattr(pytest, "pyodide_run_host_test"):
-        pytest.pyodide_run_host_test = run_host
-    if not hasattr(pytest, "pyodide_runtimes"):
-        pytest.pyodide_runtimes = runtimes
-    if not hasattr(pytest, "pyodide_dist_dir"):
-        pytest.pyodide_dist_dir = config.getoption("--dist-dir")
+    if not hasattr(pytest, "pyodide_options_stack"):
+        pytest.pyodide_options_stack = []
+    else:
+        pytest.pyodide_options_stack.append(  # type:ignore[attr-defined]
```
Comment (ryanking13) [inline pytest_pyodide/hook.py:101]: 👍

File: tests/test_doctest.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    False
+    '''
+"""
+
+ORIG_HOME = os.environ.get("HOME", None)
+
+
+def test_doctest_run(pytester, request):
+    # Help playwright find the cache
+    os.environ["XDG_CACHE_HOME"] = str(Path(ORIG_HOME) / ".cache")
```
Comment (ryanking13) [inline tests/test_doctest.py:45]: I guess Path(None) will throw error?

**Review by ryanking13 — APPROVED** [review]
Thanks @hoodmane! Codewise LGTM, but it seems like playwright is not happy for some reason.

**Comment by hoodmane** [comment]
> it seems like playwright is not happy

I tried using the selenium fixture with `getfixture` but it would require some deep pytest magic. But Joe Marshall added `get_browser_pyodide` which is very helpful here. I think the problem here is that `get_browser_pyodide` doesn't work correctly with Playwright but I'm not sure why. It also didn't shut down correctly but I think I've fixed that.

I think the correct solution is to refactor the fixtures a bit to get the selenium runtime from a separate `@cache`d function so we can also retrieve them from code. Then `get_browser_pyodide` can use that and avoid the code duplication and that will hopefully resolve the playwright problem.

**Comment by ryanking13** [comment]
Thanks for the refactorings!

About the test failure on safari, there is a known issue that safari webdriver does not allow opening multiple sessions, and we are handling it in a [very ad-hoc way](https://github.com/pyodide/pytest-pyodide/blob/2a9b65d9afe5126861e7f6badeee7a464934210c/pytest_pyodide/hook.py#L206-L221). So I am personally okay with disabling the doctest for safari if you think it is hard to handle.

**Comment by hoodmane** [comment]
Well this PR has been a lot more work than I expected. I'll give it another try and see if I can get it to work. One thing worth pointing out is that the reason this is hard is because of support for multiple runtimes. If we required only one runtime per run, there wouldn't be any difficulty at all.

**Comment by hoodmane** [comment]
@ryanking13 Could you review this again? I changed the approach quite a bit.

File: pytest_pyodide/fixture.py [hunk 3]
```diff
@@ -263,7 +263,10 @@ def selenium_context_manager(selenium_module_scope):
         selenium_module_scope.clean_logs()
         yield selenium_module_scope
     finally:
-        print(selenium_module_scope.logs)
+        try:
+            print(selenium_module_scope.logs)
+        except ValueError:
```
Comment (ryanking13) [inline pytest_pyodide/fixture.py:268]: What does this try to catch?

**Review by ryanking13 — APPROVED** [review]
Thanks @hoodmane! 

> One thing worth pointing out is that the reason this is hard is because of support for multiple runtimes. If we required only one runtime per run, there wouldn't be any difficulty at all.

If this make things too complicated, I think it's fine to only allow one runtime to be set up per run; at least within the Pyodide organization, we don't use multiple runtimes at the same time.

@joemarshall What do you think? If we go in this way, this would be a breaking change, will this cause some problem to you?

File: pytest_pyodide/fixture.py (hunk 3 above)
Comment (hoodmane) [reply pytest_pyodide/fixture.py:268]: I added a comment about it.

**Pull request merged**